src/logic/map_objects/tribes/soldier.cc

```diff
@@ -562,133 +562,143 @@ void Soldier::draw(const EditorGameBase& game,
-	// simplifies the drawing code below a lot. Before it had a lot of if () that
-	// were never tested - since our data files never changed.
-	const int dimension = attackpic->width();
-	assert(attackpic->height() == dimension);
-	assert(healthpic->width() == dimension);
-	assert(healthpic->height() == dimension);
-	assert(defensepic->width() == dimension);
-	assert(defensepic->height() == dimension);
-	assert(evadepic->width() == dimension);
-	assert(evadepic->height() == dimension);
-	assert(kSoldierHealthBarWidth > dimension);
+	{
+		// This function assumes stuff about our data files: level icons are all the
+		// same size and this is smaller than the width of the healthbar. This
+		// simplifies the drawing code below a lot. Before it had a lot of if () that
+		// were never tested - since our data files never changed.
+		const Image* healthpic = get_health_level_pic();
+
+		const Image* attackpic = get_attack_level_pic();
+		const Image* defensepic = get_defense_level_pic();
+		const Image* evadepic = get_evade_level_pic();
+
+		const int dimension = attackpic->width();
+		assert(attackpic->height() == dimension);
+		assert(healthpic->width() == dimension);
+		assert(healthpic->height() == dimension);
+		assert(defensepic->width() == dimension);
+		assert(defensepic->height() == dimension);
+		assert(evadepic->width() == dimension);
+		assert(evadepic->height() == dimension);
+		assert(kSoldierHealthBarWidth > dimension);
+	}
 #endif
 
-	const int icon_size = healthpic->width();
+	const int icon_size = get_health_level_pic()->height();
 
-	if (!anchor_below) {
+	// Draw health info in building windows, or if kSoldierLevels is on.
+	const bool draw_health_bar =
+	   draw_mode == InfoMode::kInBuilding || (info_to_draw & InfoToDraw::kSoldierLevels);
+
+	switch (draw_mode) {
+	case InfoMode::kInBuilding:
 		draw_position.x += kSoldierHealthBarWidth * scale;
 		draw_position.y += 2 * icon_size * scale;
-	} else {
-		draw_position.y -= 5 * scale;
+		break;
+	case InfoMode::kWalkingAround:
+		if (draw_health_bar) {
+			draw_position.y -= 5 * scale;
+		}
 	}
 
-	// Draw energy bar
-	assert(get_max_health());
-	const Recti energy_outer(draw_position - Vector2i(kSoldierHealthBarWidth, 0) * scale,
-	                         kSoldierHealthBarWidth * 2 * scale, 5 * scale);
-	dst->fill_rect(energy_outer, RGBColor(255, 255, 255));
+	if (draw_health_bar) {
+		// Draw energy bar
+		assert(get_max_health());
+		const RGBColor& color = owner().get_playercolor();
+		const uint16_t color_sum = color.r + color.g + color.b;
 
-	// Adjust health to current animation tick
-	uint32_t health_to_show = current_health_;
-	if (battle_) {
-		uint32_t pending_damage = battle_->get_pending_damage(this);
-		if (pending_damage > 0) {
-			int32_t timeshift = owner().egbase().get_gametime() - get_animstart();
-			timeshift = std::min(std::max(0, timeshift), 1000);
+		// The frame gets a slight tint of player color
```

Comment:
Indeed a slight tint, but pretty.

Have you tried coloring the energy_complement ? That would be better visible, but I don't know whether the current energy is discernible good enough in that case.

Reply:
I actually do, but I wanted to keep it pretty monochrome. The current formula is:
```
const RGBColor complement_color =
		   color_sum > 128 * 3 ? RGBColor(32, 32, 32) : RGBColor(224, 224, 224);
```

This was 2 years ago, so I don't really remember now how much I experimented with it at the time.